Add C++ fast path for `DTensor.__torch_dispatch__`

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #167475
* #166808
* #166372
* __->__ #167051

This patches the `__torch_dispatch__` machinery to detect DTensor and hand over control to a C++ fast path. Unlike #166370 and #166369 (which added a DTensor dispatch key and are intended to be replaced by this PR), this approach fundamentally *is* `__torch_dispatch__`, hopefully sidestepping all manner of thorny "does it work just like `__torch_dispatch__`?" that came up during development and review of #166370.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

swolchok committed:
Support C++ fast path for `DTensor.__torch_dispatch__`

This patches the `__torch_dispatch__` machinery to detect DTensor and hand over control to a C++ fast path. Unlike #166370 and #166369 (which added a DTensor dispatch key and are intended to be replaced by this PR), this approach fundamentally *is* `__torch_dispatch__`, hopefully sidestepping all manner of thorny "does it work just like `__torch_dispatch__`?" that came up during development and review of #166370.

[ghstack-poisoned]

swolchok commented:
> Lint for torch/distributed/_tools/mem_tracker.py

@sanketpurandare is working on modifying mem_tracker so it doesn't block this

swolchok committed:
forgot to reconstruct the C++ op and stack for the paths where they're not provided, fix that on "Add C++ fast path for `DTensor.__torch_dispatch__`"


This patches the `__torch_dispatch__` machinery to detect DTensor and hand over control to a C++ fast path. Unlike #166370 and #166369 (which added a DTensor dispatch key and are intended to be replaced by this PR), this approach fundamentally *is* `__torch_dispatch__`, hopefully sidestepping all manner of thorny "does it work just like `__torch_dispatch__`?" that came up during development and review of #166370.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok committed:
definitely not fully fixed, but I'm having trouble fixing the inductor_distributed failure (an undefined tensor is cropping up somewhere). let's see if we get any more specific issues out of the distributed tests on "Add C++ fast path for `DTensor.__torch_dispatch__`"


This patches the `__torch_dispatch__` machinery to detect DTensor and hand over control to a C++ fast path. Unlike #166370 and #166369 (which added a DTensor dispatch key and are intended to be replaced by this PR), this approach fundamentally *is* `__torch_dispatch__`, hopefully sidestepping all manner of thorny "does it work just like `__torch_dispatch__`?" that came up during development and review of #166370.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok committed:
fix dangling reference issue with OperatorHandle on "Add C++ fast path for `DTensor.__torch_dispatch__`"


This patches the `__torch_dispatch__` machinery to detect DTensor and hand over control to a C++ fast path. Unlike #166370 and #166369 (which added a DTensor dispatch key and are intended to be replaced by this PR), this approach fundamentally *is* `__torch_dispatch__`, hopefully sidestepping all manner of thorny "does it work just like `__torch_dispatch__`?" that came up during development and review of #166370.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok committed:
fix accidentally dispatching DTensor.__torch_function__ to __torch_dispatch__ fast path on "Add C++ fast path for `DTensor.__torch_dispatch__`"


This patches the `__torch_dispatch__` machinery to detect DTensor and hand over control to a C++ fast path. Unlike #166370 and #166369 (which added a DTensor dispatch key and are intended to be replaced by this PR), this approach fundamentally *is* `__torch_dispatch__`, hopefully sidestepping all manner of thorny "does it work just like `__torch_dispatch__`?" that came up during development and review of #166370.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok committed:
fix handling of undefined tensors coming through dispatch stack on "Add C++ fast path for `DTensor.__torch_dispatch__`"


This patches the `__torch_dispatch__` machinery to detect DTensor and hand over control to a C++ fast path. Unlike #166370 and #166369 (which added a DTensor dispatch key and are intended to be replaced by this PR), this approach fundamentally *is* `__torch_dispatch__`, hopefully sidestepping all manner of thorny "does it work just like `__torch_dispatch__`?" that came up during development and review of #166370.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok committed:
rebase, fix segfault on exit on "Add C++ fast path for `DTensor.__torch_dispatch__`"


This patches the `__torch_dispatch__` machinery to detect DTensor and hand over control to a C++ fast path. Unlike #166370 and #166369 (which added a DTensor dispatch key and are intended to be replaced by this PR), this approach fundamentally *is* `__torch_dispatch__`, hopefully sidestepping all manner of thorny "does it work just like `__torch_dispatch__`?" that came up during development and review of #166370.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok committed:
fix/support debug hooks for the sharding prop cache on "Add C++ fast path for `DTensor.__torch_dispatch__`"


This patches the `__torch_dispatch__` machinery to detect DTensor and hand over control to a C++ fast path. Unlike #166370 and #166369 (which added a DTensor dispatch key and are intended to be replaced by this PR), this approach fundamentally *is* `__torch_dispatch__`, hopefully sidestepping all manner of thorny "does it work just like `__torch_dispatch__`?" that came up during development and review of #166370.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok commented:
main distributed tests are green!

swolchok committed:
restack atop #167051 on "Add C++ fast path for `DTensor.__torch_dispatch__`"


This patches the `__torch_dispatch__` machinery to detect DTensor and hand over control to a C++ fast path. Unlike #166370 and #166369 (which added a DTensor dispatch key and are intended to be replaced by this PR), this approach fundamentally *is* `__torch_dispatch__`, hopefully sidestepping all manner of thorny "does it work just like `__torch_dispatch__`?" that came up during development and review of #166370.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

Skylion007 commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1342,6 +1697,329 @@ static PyObject* DTensor_compute_global_tensor_info(
+          break;
+        }
+        default:
+          TORCH_INTERNAL_ASSERT(false, "can't happen");
```
Apparently this does exist but it's only defined for the QNP module in PyTorch. We should generalize it and elevate it to a c10 feature.

Skylion007 reviewed (commented):
Neat: we probably should create a c10 unreachable macro.

swolchok committed:
fix handling of tensorlist arguments on "Add C++ fast path for `DTensor.__torch_dispatch__`"


This patches the `__torch_dispatch__` machinery to detect DTensor and hand over control to a C++ fast path. Unlike #166370 and #166369 (which added a DTensor dispatch key and are intended to be replaced by this PR), this approach fundamentally *is* `__torch_dispatch__`, hopefully sidestepping all manner of thorny "does it work just like `__torch_dispatch__`?" that came up during development and review of #166370.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok committed:
Update on "Add C++ fast path for `DTensor.__torch_dispatch__`"


This patches the `__torch_dispatch__` machinery to detect DTensor and hand over control to a C++ fast path. Unlike #166370 and #166369 (which added a DTensor dispatch key and are intended to be replaced by this PR), this approach fundamentally *is* `__torch_dispatch__`, hopefully sidestepping all manner of thorny "does it work just like `__torch_dispatch__`?" that came up during development and review of #166370.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

ezyang commented on torch/csrc/utils/python_arg_parser.h:
```diff
@@ -1267,6 +1267,18 @@ auto TORCH_PYTHON_API handle_torch_function_no_python_arg_parser(
+    PyObject* torch_api_function,
+    const char* module_name,
+    const c10::OperatorHandle* opt_op,
+    torch::jit::Stack* opt_stack,
```
These two args are the difference here

swolchok committed:
rebase to main to pick up #167580 on "Add C++ fast path for `DTensor.__torch_dispatch__`"


This patches the `__torch_dispatch__` machinery to detect DTensor and hand over control to a C++ fast path. Unlike #166370 and #166369 (which added a DTensor dispatch key and are intended to be replaced by this PR), this approach fundamentally *is* `__torch_dispatch__`, hopefully sidestepping all manner of thorny "does it work just like `__torch_dispatch__`?" that came up during development and review of #166370.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1342,6 +1697,329 @@ static PyObject* DTensor_compute_global_tensor_info(
+  return contains_any_symint(shape);
+}
+
+static std::optional<NativeOpSchema> create_native_op_schema(
```
So, this isn't actually fast path, but we need to somehow get the op schema into C++ land I guess?

ezyang replied:
No, it is in fact fast path (always called in the dispatch)

swolchok replied:
this is the fast path. we need a key for the sharding propagator cache. (it would be entirely reasonable to propose that we rename NativeOpSchema to NativeShardingPropagatorCache key, I think)

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1010,6 +1091,268 @@ static c10::SymDimVector tuple_to_symintlist(PyObject* obj) {
+  all_thread_caches.clear();
+}
+
+py::object dispatchDTensorOp(
```
The main entry point!!

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1342,6 +1697,329 @@ static PyObject* DTensor_compute_global_tensor_info(
+              // WARNING: here we rely on c10::List being represented
+              // by a contiguous array of IValue for efficiency!
+              arg = c10::ivalue::Tuple::create(c10::ArrayRef<c10::IValue>(
+                  &(*list.begin()).get(), list.size()));
```
Yes, I think it actually wouldn't be horrible to handle Tensor here, but since we excluded the pytree ones you don't hit that here.

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1342,6 +1697,329 @@ static PyObject* DTensor_compute_global_tensor_info(
+
+  if (native_info.static_kwargkey && !native_info.static_kwargkey.is_none()) {
+    // Separator to disambiguate kwargs from args in comparison and hashing.
+    static constexpr int64_t kwargs_separator = 0x0011223344556677LL;
```
magic number :suspect:

ezyang reviewed (approved):
This feels good, thank you

swolchok committed:
rebase to viable/strict, fix some tail build issues and add a couple comments on "Add C++ fast path for `DTensor.__torch_dispatch__`"


This patches the `__torch_dispatch__` machinery to detect DTensor and hand over control to a C++ fast path. Unlike #166370 and #166369 (which added a DTensor dispatch key and are intended to be replaced by this PR), this approach fundamentally *is* `__torch_dispatch__`, hopefully sidestepping all manner of thorny "does it work just like `__torch_dispatch__`?" that came up during development and review of #166370.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok commented:
Hmm, the macos-py3-arm64 test failures look real (`ModuleNotFoundError: No module named 'torch._C._distributed_c10d'; 'torch._C' is not a package`, doesn't seem to be firing in HUD) but they don't go through anything I changed. attempting to debug locally.

swolchok commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -301,6 +303,12 @@ static py::object maybe_get_registered_torch_dispatch_rule(
 }
 
+static bool is_dtensor(PyObject* obj) {
+  const py::handle dtensor = get_dtensor_class();
```
ezyang says gate with ifdef USE_DISTRIBUTED

swolchok committed:
fix mac tests at bottom of stack with USE_DISTRIBUTED gating in is_dtensor on "Add C++ fast path for `DTensor.__torch_dispatch__`"


This patches the `__torch_dispatch__` machinery to detect DTensor and hand over control to a C++ fast path. Unlike #166370 and #166369 (which added a DTensor dispatch key and are intended to be replaced by this PR), this approach fundamentally *is* `__torch_dispatch__`, hopefully sidestepping all manner of thorny "does it work just like `__torch_dispatch__`?" that came up during development and review of #166370.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok committed:
fix mac tests at bottom of stack, second try on "Add C++ fast path for `DTensor.__torch_dispatch__`"


This patches the `__torch_dispatch__` machinery to detect DTensor and hand over control to a C++ fast path. Unlike #166370 and #166369 (which added a DTensor dispatch key and are intended to be replaced by this PR), this approach fundamentally *is* `__torch_dispatch__`, hopefully sidestepping all manner of thorny "does it work just like `__torch_dispatch__`?" that came up during development and review of #166370.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]
